test/test_jit_cuda_fuser.py

```diff
@@ -4233,6 +4233,59 @@ def t(x_kj: torch.Tensor, idx_kj: torch.Tensor, sbf: torch.Tensor, dim : int):
         t_jit = torch.jit.script(t)
         self._run_helper(t_jit, t, lookup_tv, indies_tv, sbf, dim)
 
+    @unittest.skipIf(not RUN_NVFUSER, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_gather_backward(self):
+        lookup_size = 68
+        feat_dim = 128
+        select_dim = 1
+
+        x = torch.randn(lookup_size, feat_dim, dtype=torch.float, device="cuda") .requires_grad_()
+        y = torch.randint(0, lookup_size, (lookup_size, select_dim), device="cuda").to(dtype=torch.long)
+        z = torch.rand(lookup_size, select_dim, dtype=torch.float, device="cuda").requires_grad_()
+        grad = torch.randn(lookup_size, select_dim, dtype=torch.float, device="cuda")
+
+        def t(x, y, z):
+            o = torch.mul(x, z)
+            o = torch.gather(o, 1, y)
+            return o
+        t_jit = torch.jit.script(t)
+        self._run_training_helper(t_jit, t, grad, x, y, z)
+
+    @unittest.skipIf(not RUN_NVFUSER, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_gather_fusion(self):
+        lookup_size = 68
+        feat_dim = 128
+        select_dim = 1
+
+        x = torch.randn(lookup_size, feat_dim, dtype=torch.float, device="cuda") 
+        y = torch.randint(0, lookup_size, (lookup_size, select_dim), device="cuda").to(dtype=torch.long)
+        z = torch.rand(lookup_size, select_dim, dtype=torch.float, device="cuda")
+
+        def t(x, y, z):
+            o = torch.gather(x, 1, y) * z + 176
+            return o
+        t_jit = torch.jit.script(t)
+        self._run_helper(t_jit, t, x, y, z)
+
+    @unittest.skipIf(not RUN_NVFUSER, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_gather_sparse_grad(self):
+        x = torch.randn(2, 2, dtype=torch.float, device="cuda").requires_grad_()
+        y = torch.randint(0, 1, (2, 2), device="cuda").to(dtype=torch.long)
+        z = torch.rand(2, 2, dtype=torch.float, device="cuda").requires_grad_()
+        grad = torch.randn(2, 2, dtype=torch.float, device="cuda")
+        def t(x, y, z):
+            o = torch.gather(x + z, 1, y + y, sparse_grad=True)
+            return o
+
+        t_jit = torch.jit.script(t)
+        self._run_training_helper(t_jit, t, grad, x, y, z)
+
```

Comment:
Actually, `test_gather_sparse_grad` will raise `graph contains 0 prim::CudaFusionGuard nodes (including subgraphs) but expected 1` error. But I do not know how to fix it.

Reply:
It should be failing here: https://github.com/csarofeen/pytorch/blob/1f39c3a1cacd1e83ab7dd519931076567fe81626/test/test_jit_cuda_fuser.py#L252

Since we don't have a backward rule for gather yet.
We can add two more default arguments to the function. `num_fw_fusion=1, num_bw_fusion=1` there, similar to what we did with `run_helper`: https://github.com/csarofeen/pytorch/blob/1f39c3a1cacd1e83ab7dd519931076567fe81626/test/test_jit_cuda_fuser.py#L197

Thanks in advance for doing the clean up :stuck_out_tongue_winking_eye: